build_runner/lib/src/asset_graph/graph.dart

```diff
@@ -262,23 +272,33 @@ class AssetGraph {
 
   /// Adds [outputs] as [GeneratedAssetNode]s to the graph.
   ///
-  /// Replaces any existing [AssetNode]s with [GeneratedAssetNode]s.
-  void _addGeneratedOutputs(Iterable<AssetId> outputs, int phaseNumber,
+  /// If there are existing [SourceAssetNode]s or [SyntheticAssetNode]s  that
+  /// overlap the [GeneratedAssetNode]s, then they will be replaced with
+  /// [GeneratedAssetNode]s, and all their `primaryOutputs` will be removed
+  /// from the graph as well. The return value is the set of assets that were
+  /// removed from the graph.
+  Set<AssetId> _addGeneratedOutputs(Iterable<AssetId> outputs, int phaseNumber,
       {AssetId primaryInput}) {
+    var removed = new Set<AssetId>();
     for (var output in outputs) {
       // When `writeToCache` is false we can pick up old generated outputs as
-      // regular `AssetNode`s, this deletes them.
-      if (contains(output)) _remove(output);
+      // regular `AssetNode`s, we need to delete them and all their primary
+      // outputs, and replace them with a `GeneratedAssetNode`.
+      if (contains(output)) {
+        assert(get(output) is! GeneratedAssetNode);
+        _remove(output, removedSet: removed);
+      }
 
       _add(new GeneratedAssetNode(
           phaseNumber, primaryInput, true, false, output));
     }
+    return removed;
   }
 
   @override
   String toString() => allNodes.toList().toString();
 
   // TODO remove once tests are updated
   void add(AssetNode node) => _add(node);
-  AssetNode remove(AssetId id) => _remove(id);
+  Set<AssetId> remove(AssetId id) => _remove(id);
```

Comment:
Are we going to have to keep this? May as well drop the private version if we have an exact forwarding public method.

Reply:
See the TODO above, at some point I would like to delete these, it just isn't a priority right now.